### PR TITLE
Use sub group reduce operation in warp reduce.

### PR DESCRIPTION
**Pull request by chengjunlu**
Use the Triton Gen subgroup reduce operator for the warp reduce.

### CODE REVIEW

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 1]
```diff
@@ -1036,6 +1036,9 @@ struct TritonSubGroupReduceLowering
     LLVM::LLVMFuncOp funcOp =
         LLVM::lookupOrCreateFn(moduleOp, funcName, argTypes, val_ty);
     funcOp.setCConv(LLVM::cconv::CConv::SPIR_FUNC);
+    auto convergentAttr =
+        rewriter.getArrayAttr(StringAttr::get(op->getContext(), "convergent"));
+    funcOp.setPassthroughAttr(convergentAttr);
 
     auto callOp = rewriter.create<LLVM::CallOp>(loc, funcOp, args);
     callOp.setCConv(LLVM::cconv::CConv::SPIR_FUNC);
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1044]: Attributes on function and call op should match, can you please add convergent to the call op as well?

File: third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp [hunk 2]
```diff
@@ -86,8 +86,66 @@ bool TargetInfo::warpReduce(ConversionPatternRewriter &rewriter, Location loc,
                             SmallVector<Value> &acc, triton::ReduceOp op,
                             unsigned numLaneToReduce,
                             unsigned interleave) const {
-  assert("TODO: implement warpReduce on XPU");
-  return false;
+  // No horizontal reduce required.
+  if (numLaneToReduce == 1)
+    return false;
+  // horizontal reduce with interleave stride not support.
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp:92]:
```suggestion
  // Horizontal reduce with interleave stride not support.
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp [hunk 3]
```diff
@@ -86,8 +86,66 @@ bool TargetInfo::warpReduce(ConversionPatternRewriter &rewriter, Location loc,
...
+  Operation *reduceOp = yield->getOperand(0).getDefiningOp();
+  if (!reduceOp || reduceOp->getNumOperands() != 2 ||
+      reduceOp->getNumResults() != 1)
+    return false;
+  if (reduceOp->getOperand(0) != block.getArgument(0) ||
+      reduceOp->getOperand(1) != block.getArgument(1))
+    return false;
+
+  std::optional<TritonGEN::ReduceKind> reduceKind = std::nullopt;
+  if (isa<arith::AddFOp>(reduceOp))
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp:112]: can use TypeSwitch?

File: test/Conversion/intel/tritongpu_to_gen.mlir [hunk 4]
```diff
@@ -1370,6 +1370,275 @@ module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 1 :
...
+    // CHECK: %[[HELPER_LANE:.*]] = llvm.mlir.constant(0 : i32) : i32
+    // CHECK: {{.*}} = llvm.call spir_funccc @llvm.genx.GenISA.WaveAll.i1({{.*}}, %[[REDUCE_KIND]], %[[HELPER_LANE]]) : (i1, i8, i32) -> i1
+    %0 = "tt.reduce"(%arg) <{axis = 0 : i32}> ({
+    ^bb0(%arg4: i1, %arg5: i1):
+      %48 = arith.addi %arg4, %arg5 : i1
+      tt.reduce.return %48 : i1
+    }) : (tensor<256x1xi1, #blocked>) -> tensor<1xi1, #slice>
+
+    // CHECK: %[[REDUCE_KIND:.*]] = llvm.mlir.constant(8 : i8) : i8
+    // CHECK: %[[HELPER_LANE:.*]] = llvm.mlir.constant(0 : i32) : i32
```
Comment (whitneywhtsang) [inline test/Conversion/intel/tritongpu_to_gen.mlir:1389]: missing a check to WaveAll here?

File: third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp (hunk 3 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp]: Yes

File: test/Conversion/intel/tritongpu_to_gen.mlir (hunk 4 above)
Comment (chengjunlu) [reply test/Conversion/intel/tritongpu_to_gen.mlir:1389]: Added

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 1 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1044]: Let's use `createDeviceFunctionCall` and the new Attribute builder facilities to create all of the device function please.

**Review by etiotto — DISMISSED** [review]

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 1 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1044]: why did you resolve this comment without addressing it ? Reopening, please address the comment. Thanks.

**Review by etiotto — CHANGES_REQUESTED** [review]

File: third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp [hunk 5]
```diff
@@ -86,8 +87,67 @@ bool TargetInfo::warpReduce(ConversionPatternRewriter &rewriter, Location loc,
...
+      .Case<arith::MinSIOp>(
+          [&](auto) { reduceKind = TritonGEN::ReduceKind::IMIN; })
+      .Case<arith::MinUIOp>(
+          [&](auto) { reduceKind = TritonGEN::ReduceKind::UMIN; })
+      .Case<arith::MaxSIOp>(
+          [&](auto) { reduceKind = TritonGEN::ReduceKind::IMAX; })
+      .Case<arith::MaxUIOp>(
+          [&](auto) { reduceKind = TritonGEN::ReduceKind::UMAX; })
+      .Case<arith::MulIOp>(
+          [&](auto) { reduceKind = TritonGEN::ReduceKind::PROD; });
```
Comment (FMarno) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp:139]:
```suggestion
```suggestion
    std::optional<TritonGEN::ReduceKind> reduceKind =  llvm::TypeSwitch<mlir::Operation *, TritonGEN::ReduceKind>(reduceOp)
      .Case<arith::AddFOp>(
          []{ return TritonGEN::ReduceKind::FSUM; })
      .Case<arith::MaxNumFOp>(
          [] { return TritonGEN::ReduceKind::FMAX; })
      .Case<arith::MinNumFOp>(
          [] { return TritonGEN::ReduceKind::FMIN; })
      .Case<arith::MulFOp>(
          [] { return TritonGEN::ReduceKind::FPROD; })
      .Case<arith::AddIOp>(
          [] { return TritonGEN::ReduceKind::SUM; })
      .Case<arith::AndIOp>(
          [] { return TritonGEN::ReduceKind::AND; })
      .Case<arith::OrIOp>([] { return TritonGEN::ReduceKind::OR; })
      .Case<arith::XOrIOp>(
          [] { return TritonGEN::ReduceKind::XOR; })
      .Case<arith::MinSIOp>(
          [] { return TritonGEN::ReduceKind::IMIN; })
      .Case<arith::MinUIOp>(
          [] { return TritonGEN::ReduceKind::UMIN; })
      .Case<arith::MaxSIOp>(
          [] { return TritonGEN::ReduceKind::IMAX; })
      .Case<arith::MaxUIOp>(
          [] { return TritonGEN::ReduceKind::UMAX; })
      .Case<arith::MulIOp>(
          [] { return TritonGEN::ReduceKind::PROD; })
       .Default(std::nullopt);
```
```

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp]: Done in https://github.com/intel/intel-xpu-backend-for-triton/pull/1277.

File: third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp (hunk 5 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp]: done.

**Review by whitneywhtsang — APPROVED** [review]
@chengjunlu Please confirm you are ok with the new changes.

**Comment by chengjunlu** [comment]
> @chengjunlu Please confirm you are ok with the new changes.

Yes. LGTM

File: third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp [hunk 6]
```diff
@@ -86,8 +87,53 @@ bool TargetInfo::warpReduce(ConversionPatternRewriter &rewriter, Location loc,
...
+    return false;
+  // Check if it is a simple reduce operation supported by Wave Op.
+  if (op.getNumOperands() != 1 || op.getNumResults() != 1)
+    return false;
+  auto &combineOp = op.getCombineOp();
+  if (combineOp.getBlocks().size() > 1)
+    return false;
+  Block &block = *combineOp.begin();
+  Operation *yield = block.getTerminator();
+  Operation *reduceOp = yield->getOperand(0).getDefiningOp();
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp:104]: Is it always true that the first operand of the terminator instruction is a reduce operation in this context?
The variable name imply so. If this is the case one should add an assert. It it is not always true the variable name should be generic.

**Pull request merged**